File: app/src/main/java/com/anthony/deepl/openl/model/TranslationResponse.java
package com.anthony.deepl.openl.model;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;

public class TranslationResponse {

    @SerializedName("id")
    private String mId;

    @SerializedName("jsonrpc")
    private String mJsonRpc;

    @SerializedName("result")
    private TranslationResponseResult mResult;

    public String getSourceLanguage() {
        return mResult.getSourceLanguage();
    }

    public String getBestTranslation(List<Integer> lineBreakPositions) {
        String bestTranslation = "";
        List<TranslationResponseTranslations> translations = mResult.getTranslations();
        if (lineBreakPositions == null) {
            lineBreakPositions = new ArrayList<>();
        }

        for (int i = 0, translationSize = translations != null ? translations.size() : 0; i < translationSize; i++) {
            String bestResultSentence = null;
            Float bestResultScore = null;
            List<TranslationResponseBeam> beams = translations.get(i).getBeams();
            for (int j = 0, size = beams != null ? beams.size() : 0; j < size; j++) {
                TranslationResponseBeam beam = beams.get(j);
                if (bestResultScore == null || bestResultScore < beam.getTotalProbability()) {
                    bestResultScore = beam.getTotalProbability();
                    bestResultSentence = beam.getTranslatedSentence();
                }
            }
            if (bestResultSentence != null) {
                bestTranslation = bestTranslation.concat(bestResultSentence);
            }
            if (i + 1 < translationSize) {
                bestTranslation = bestTranslation.concat(lineBreakPositions.contains(i) ? "\n" : " ");
            }
        }
        return bestTranslation;
    }

    public List<String> getOtherResults() {
        if (mResult.getTranslations() == null || mResult.getTranslations().size() > 1) {
            return null;
        }
        int bestResultPosition = -1;
        Float bestResultScore = null;
        List<String> otherSentences = new ArrayList<>();
        List<TranslationResponseBeam> beams = mResult.getTranslations().get(0).getBeams();

        for (int i = 0, size = beams != null ? beams.size() : 0; i < size; i++) {
            TranslationResponseBeam beam = beams.get(i);
            otherSentences.add(beam.getTranslatedSentence());
            if (bestResultScore == null || bestResultScore < beam.getTotalProbability()) {
                bestResultScore = beam.getTotalProbability();
                bestResultPosition = i;
            }
        }

        if (bestResultPosition != -1 && otherSentences.size() > bestResultPosition) {
            otherSentences.remove(bestResultPosition);
        }

        return otherSentences;
    }

}

class TranslationResponseResult {

    @SerializedName("source_lang")
    private String mSourceLanguage;

    @SerializedName("source_lang_is_confident")
    private Float mSourceLanguageConfinence;

    @SerializedName("target_lang")
    private String mTargetLanguage;

    @SerializedName("translations")
    private List<TranslationResponseTranslations> mTranslations;

    public String getSourceLanguage() {
        return mSourceLanguage;
    }

    public List<TranslationResponseTranslations> getTranslations() {
        return mTranslations;
    }

}

class TranslationResponseTranslations {

    @SerializedName("timeAfterPreprocessing")
    private Double mAfterPreprocessingTime;

    @SerializedName("timeReceivedFromEndpoint")
    private String mReceivedFromEndpointTime;

    @SerializedName("timeSentToEndpoint")
    private Float mSendToEndpointTime;

    @SerializedName("total_time_endpoint")
    private Float mTotalEndpointTime;

    @SerializedName("beams")
    private List<TranslationResponseBeam> mBeams;

    public List<TranslationResponseBeam> getBeams() {
        return mBeams;
    }

}

class TranslationResponseBeam {

    @SerializedName("num_symbols")
    private int mTranslationLength;

    @SerializedName("postprocessed_sentence")
    private String mTranslatedSentence;

    @SerializedName("score")
    private Float mTranslationScore;

    @SerializedName("totalLogProb")
    private Float mTotalProbability;

    public float getTranslationScore() {
        return mTranslationScore;
    }

    public float getTotalProbability() {
        return mTotalProbability;
    }

    public String getTranslatedSentence() {
        return mTranslatedSentence;
    }

}

File: app/src/main/java/com/anthony/deepl/openl/model/TranslationResponse.kt
package com.anthony.deepl.openl.model

import com.squareup.moshi.Json
import java.util.ArrayList

data class TranslationResponse (
        @field:Json(name="id") val id: String?,
        @field:Json(name="jsonrpc") val jsonRpc: String?) {

    @field:Json(name="result")
    private val mResult: TranslationResponseResult? = null
    var lineBreakPositions: MutableList<Int> = mutableListOf()

    val sourceLanguage: String?
        get() = mResult?.sourceLanguage

    fun getBestTranslation(): String {
        var bestTranslation = ""

        mResult?.translations?.forEachIndexed { index, sentence ->
            var bestResultSentence: String? = null
            var bestResultScore: Float = -1F
            sentence.beams?.forEach {
                if (bestResultScore == -1F || bestResultScore < it.totalProbability) {
                    bestResultScore = it.totalProbability
                    bestResultSentence = it.translatedSentence
                }
            }
            if (bestResultSentence != null) {
                bestTranslation += bestResultSentence
                bestTranslation += if (lineBreakPositions.contains(index) == true) "\n" else " "
            }
        }

        return bestTranslation
    }

    fun getAlternateTranslations(): List<String>? {
        if (mResult?.translations?.size ?: 0 > 1) {
            return null
        }
        var bestResultPosition = -1
        var bestResultScore: Float = -1F
        val otherSentences = ArrayList<String>()

        mResult?.translations?.get(0)?.beams?.forEachIndexed { index, beam ->
            otherSentences.add(beam.translatedSentence)
            if (bestResultScore == -1F || bestResultScore < beam.totalProbability) {
                bestResultScore = beam.totalProbability
                bestResultPosition = index
            }
        }

        if (bestResultPosition != -1 && otherSentences.size > bestResultPosition) {
            otherSentences.removeAt(bestResultPosition)
        }

        return otherSentences
    }
}

internal data class TranslationResponseResult(
        @field:Json(name="source_lang") val sourceLanguage: String?,
        @field:Json(name="source_lang_is_confident") val sourceLanguageConfidence: Float?,
        @field:Json(name="target_lang") val targetLanguage: String?,
        @field:Json(name="translations") val translations: List<TranslationResponseTranslations>?
)

internal data class TranslationResponseTranslations(
        @field:Json(name="timeAfterPreprocessing") val afterPreprocessingTime: Double?,
        @field:Json(name="timeReceivedFromEndpoint") val receivedFromEndpointTime: String?,
        @field:Json(name="timeSentToEndpoint") val sendToEndpointTime: Float?,
        @field:Json(name="total_time_endpoint") val mTotalEndpointTime: Float?,
        @field:Json(name="beams") val beams: List<TranslationResponseBeam>?
)

internal data class TranslationResponseBeam(
        @field:Json(name="num_symbols") val translationLength: Int,
        @field:Json(name="postprocessed_sentence") val translatedSentence: String,
        @field:Json(name="score") val translationScore: Float?,
        @field:Json(name="totalLogProb") val totalProbability: Float
)